ntp-daemon/src/ipfilter.rs

```diff
@@ -0,0 +1,328 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+use crate::config::subnet::IpSubnet;
+
+/// One part of a BitTree
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+struct Treenode {
+    // Where in the array the child nodes of this
+    // node are located. A child node is only
+    // generated if the symbol cannot be used to
+    // make a final decision at this level
+    child_offset: u32,
+    inset: u16,
+    outset: u16,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// BitTree is a Trie on 128 bit integers encoding
+/// which integers are part of the set.
+///
+/// It matches the integer a 4-bit segment at a time
+/// recording at each level whether for a given symbol
+/// all integers with the prefix extended with that
+/// symbol are either in or outside of the set.
+struct BitTree {
+    nodes: Vec<Treenode>,
+}
+
+fn top_quad(v: u128) -> u8 {
+    ((v & (0xF << 124)) >> 124) as u8
```

Comment:
does `(v >> 124) & 0b1111` have problems? saves an instruction (or multiple because 128 bits do not fit in a single register)

also, why `quad` in the name? 4 bits are usually called a [nibble](https://en.wikipedia.org/wiki/Nibble)

Reply:
fixed